### PR TITLE
BUG: Fix performance issue when copying vtkMRMLStreamingVolumeNode

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
Copying a vtkMRMLStreamingVolumeNode would cause the frame to be decoded, leading to to poor performance when copying entire sequences.
Fixed by preventing decoding during node copy.

### CODE REVIEW

**Comment by jcfr** [comment]
The change makes sense, I am wondering if we should take this opportunity to add the `CopyInProgress` ivar to the base class and update the function `CopyContent`

@lassoan What do you think ?

**Comment by lassoan** [comment]
> The change makes sense, I am wondering if we should take this opportunity to add the CopyInProgress ivar to the base class and update the function CopyContent

I think this is a special case and it should not be necessary in general. Have you seen similar flags in other classes?

**Comment by Sunderlandkyl** [comment]
Are there any changes needed, or can this be integrated?

**Comment by lassoan** [comment]
Calling the Superclass::CopyContent with some special flags make things a bit opaque, especially because vtkMRMLVolumeNode::CopyContent is already quite complicated (it copies image data, emits image modified events, etc.).

It would be simpler, more explicit if we did not call Superclass::CopyContent in vtkMRMLStreamingVolumeNode::CopyContent but call vtkMRMLDisplayableNode::CopyContent instead and then implement all necessary volume copying operations there. It would result in some code duplication (vtkMRMLTensorVolumeNode::CopyContent and parts of vtkMRMLVolumeNode::CopyContent), but it is just a couple of lines, and so it should not be a problem.

**Comment by Sunderlandkyl** [comment]
@lassoan Ok, I've duplicated some of the CopyContent code from vtkMRMLTensorVolumeNode and vtkMRMLVolumeNode in a way that avoid decoding the frame of the source node.

File: Libs/MRML/Core/vtkMRMLStreamingVolumeNode.cxx [hunk 1]
```diff
@@ -340,16 +340,45 @@ void vtkMRMLStreamingVolumeNode::ReadXMLAttributes(const char** atts)
...
+      producer ? producer->GetOutputDataObject(
+        this->ImageDataConnection->GetIndex()) : nullptr);
+
+    vtkSmartPointer<vtkImageData> targetImageData = sourceImageData;
+    if (deepCopy && sourceImageData)
+      {
+      targetImageData = vtkSmartPointer<vtkImageData>::Take(sourceImageData->NewInstance());
+      targetImageData->DeepCopy(sourceImageData);
+      }
+      this->SetAndObserveImageData(targetImageData);
```
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLStreamingVolumeNode.cxx:368]: indentation

File: Libs/MRML/Core/vtkMRMLStreamingVolumeNode.cxx [hunk 2]
```diff
@@ -340,16 +340,45 @@ void vtkMRMLStreamingVolumeNode::ReadXMLAttributes(const char** atts)
...
+  if (!streamingVolumeNode)
     {
-    return;
+    // Other node is not a streaming volume node.
+    // We don't need to worry about it trying to decode the frame.
+    Superclass::CopyContent(anode, deepCopy);
+    }
+  else
+    {
+    vtkMRMLDisplayableNode::CopyContent(anode, deepCopy);
```
Comment (lassoan) [inline Libs/MRML/Core/vtkMRMLStreamingVolumeNode.cxx:353]: Describe here why we don't call SuperClass::CopyContent.

**Review by lassoan — COMMENTED** [review]
Thank you, this looks good, clean, just address the two trivial comments.

**Comment by Sunderlandkyl** [comment]
Thanks, changes made.

**Review by lassoan — APPROVED** [review]
Thank you, looks good.

**Pull request merged**